synapse/common.py

```diff
@@ -1238,6 +1238,30 @@ async def wait_for(fut, timeout):
     async with _timeout(timeout):
         return await fut
 
+async def waitretn(futu, timeout):
+    try:
+        valu = await wait_for(futu, timeout)
+        return (True, valu)
+    except Exception as e:
+        return (False, excinfo(e))
+
+async def waitgenr(genr, timeout):
+
+    genr = genr.__aiter__()
+    try:
```

Comment:
This is pretty straightforward function so it probably doesn't matter but you could use `contextlib.aclosing` here instead to ensure the genr is cleaned up.